packages-rs/drainpipe/src/main.rs

```diff
@@ -10,16 +10,28 @@ mod firehose;
 mod schema;
 
 #[derive(Debug)]
-enum ProcessError {
-    DecodeError { _e: firehose::Error },
-    ProcessError { seq: i64, error: anyhow::Error },
+enum ProcessErrorKind {
+    DecodeError,
+    ProcessError,
+}
+
+#[derive(Debug)]
+struct ProcessError {
+    seq: i64,
+    inner: anyhow::Error,
+    msg: Vec<u8>,
+    kind: ProcessErrorKind,
 }
 
 /// Process a message from the firehose. Returns the sequence number of the message or an error.
 async fn process(message: Vec<u8>, ctx: &mut Context) -> Result<i64, ProcessError> {
-    let (_header, message) =
-        firehose::read(&message).map_err(|e| ProcessError::DecodeError { _e: e })?;
-    let sequence = match message {
+    let (_header, data) = firehose::read(&message).map_err(|e| ProcessError {
+        inner: e.into(),
+        seq: -1,
+        msg: message.clone(),
+        kind: ProcessErrorKind::DecodeError,
+    })?;
+    let sequence = match data {
```

Comment:
renamed `message` here to `data`, to keep the reference to the original message (rather than the data read from and returned by `firehose::read`) in order to toss it into the potential error downstream